src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,319 @@
+
+        # Set the JOB_DIR
+        os.environ["JOB_DIR"] = self.job_dir
+
+    def stop(self):
+        loader_helper.stop_podman()
+
+
+class Conf:
+    def __init__(self, job_dir, output_dir, workers):
+        self.cb = Callback()
+        self.cb.start()
+        time.sleep(2)
+        # os.environ['SDK_CALLBACK_URL'] = self.cb.callback_url
+        token = os.environ["KB_AUTH_TOKEN"]
+        self.ws = Workspace(WS_DOMAIN, token=token)
+        self.asu = AssemblyUtil(self.cb.callback_url, token=token)
+        self.queue = Queue()
+        self.pth = output_dir
+        self.job_dir = job_dir
+        self.pools = Pool(workers, process_input, [self])
+
+
+def _make_output_dir(root_dir, source_data_dir, source):
+    """Helper function that makes working directory for a specific collection under root directory"""
+
+    if source == "WS":
+        return os.path.join(root_dir, source_data_dir, source)
+    raise ValueError(f"Unexpected source: {source}")
+
+
+def _make_job_dir(project_dir, username):
+    """Helper function that create a job dir if not provided by the user"""
+    return os.path.join(project_dir, username)
+
+
+def _list_objects_params(wsid, min_id, max_id):
+    """Helper function that creats params needed for list_objects function"""
+    params = {"ids": [wsid], "minObjectID": min_id, "maxObjectID": max_id}
+    return params
+
+
+def _process_object_info(obj_info):
+    """
+    "upa", "name", "type", and "timestamp info will be extracted from object info and save as a dict"
+    {
+        "upa": "790541/67/2",
+        "name": <copy object name from object info>
+        "type": <copy object type from object info>
+        "timestamp": <copy timestamp from object info>
+    }
+    """
+    res_dict = {}
+    res_dict["upa"] = "{6}/{0}/{4}".format(*obj_info)
+    res_dict["name"] = obj_info[1]
+    res_dict["type"] = obj_info[2]
+    res_dict["timestamp"] = obj_info[3]
+    return res_dict
+
+
+def list_objects(wsid, conf, batch_size, filter_objects_name_by=None):
+    """
+    List all objects information given a workspace ID
+    """
+
+    maxObjectID = conf.ws.get_workspace_info({"id": wsid})[4]
+    batch_input = [
+        [idx + 1, idx + batch_size] for idx in range(0, maxObjectID, batch_size)
+    ]
+    objs = [
+        conf.ws.list_objects(_list_objects_params(wsid, min_id, max_id))
+        for min_id, max_id in batch_input
+    ]
+    res_objs = list(itertools.chain.from_iterable(objs))
+    if filter_objects_name_by:
+        res_objs = [
+            obj for obj in res_objs if obj[2].startswith(filter_objects_name_by)
+        ]
+    return res_objs
+
+
+def process_input(conf):
+    """
+    Download .fa and .meta files from workspace and save a copy under output_dir
+    """
+    while True:
+        upa, obj_info = conf.queue.get(block=True)
+        if not upa:
+            print("Stopping")
+            break
+        cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
```

Comment:
should `job_dir` be deleted after the whole process finished?

Reply:
maybe with a CLI flag to keep it for debugging purposes

Reply:
delete_job_dir flag is added.

Reply:
:+1: